MetadataExtractor/IO/ReaderInfo.cs

```diff
@@ -0,0 +1,694 @@
+        }
+
+        private long GlobalPosition => StartPosition + LocalPosition;
+        public long StartPosition { get; private set; }
+        public long LocalPosition { get; private set; }
+
+        public long Length
+        {
+            get
+            {
+                return (p_length != -1) ? p_length : (p_ras.Length - StartPosition);
+            }
+        }
+
+        /// <summary>Get and set the byte order of this reader. <c>true</c> by default.</summary>
+        /// <remarks>
+        /// <list type="bullet">
+        ///   <item><c>true</c> for Motorola (or big) endianness (also known as network byte order), with MSB before LSB.</item>
+        ///   <item><c>false</c> for Intel (or little) endianness, with LSB before MSB.</item>
+        /// </list>
+        /// </remarks>
+        /// <value><c>true</c> for Motorola/big endian, <c>false</c> for Intel/little endian</value>
+        public bool IsMotorolaByteOrder { get; set; }
+
+        /// <summary>
+        /// Creates a new <see cref="ReaderInfo"/> with the current properties of this reader
+        /// </summary>
+        /// <returns></returns>
+        public ReaderInfo Clone() => Clone(0, -1, true);
+        public ReaderInfo Clone(bool useByteOrder) => Clone(0, useByteOrder);
+        public ReaderInfo Clone(long length) => Clone(0, length, true);
+        public ReaderInfo Clone(long offset, long length) => Clone(offset, length, true);
+        public ReaderInfo Clone(long offset, bool useByteOrder) => Clone(offset, -1, useByteOrder);
+        public ReaderInfo Clone(long offset, long length, bool useByteOrder)
+        {
+            //return p_ras.CreateReader(GlobalPosition + offset, (length > -1 ? length : Length), useByteOrder ? IsMotorolaByteOrder : !IsMotorolaByteOrder);
+            return p_ras.CreateReader(GlobalPosition + offset, (length > -1 ? length : Length - offset), useByteOrder ? IsMotorolaByteOrder : !IsMotorolaByteOrder);
+        }
+
+        public static ReaderInfo CreateFromArray(byte[] bytes)
+        {
+            //if (bytes == null)
+            //    throw new ArgumentNullException(nameof(bytes));
+
+            return new RandomAccessStream(bytes).CreateReader();
+        }
+
+        public static ReaderInfo CreateFromStream(Stream stream, long streamLength = -1)
+        {
+            //if (bytes == null)
+            //    throw new ArgumentNullException(nameof(bytes));
+
+            return new RandomAccessStream(stream, streamLength).CreateReader();
+        }
+
+        /// <summary>Seeks forward or backward in the sequence.</summary>
+        /// <remarks>
+        /// Skips forward or backward in the sequence. If the sequence ends, an <see cref="IOException"/> is thrown.
+        /// </remarks>
+        /// <param name="offset">the number of bytes to seek, in either direction.</param>
+        /// <param name="isSequential">optional for testing to indicate whether sequential access was used.</param>
```

Comment:
I'm unclear on why `isSequential` is needed for these `Skip` methods.

Reply:
SequentialReader classes throw an IOException when too many bytes are requested. IndexedReader classes throw a BufferBoundsException. It was added very recently as a flag to help throw different exceptions to match your original readers, and make Test methods match better.

It isn't necessary at all if you want it to throw a common exception.

Reply:
Thanks. I think throwing `BufferBoundsException` everywhere would be fine. It derives from `IOException` anyway, and includes more information. This parameter seems out of place somehow, and it gets passed around a lot which will have some perf impact.

It looks like it's controlling some logic in `ReadAtGlobal` though.

I'm a bit unclear on how `ReaderInfo` functions in both sequential and indexed access. Can you switch between them for the same instance at leisure? What is the rationale for having a single type versus two different types?

Reply:
A ReaderInfo class can be read indexed or sequentially at the same time. You are free to make indexed calls at any time and only sequential calls will advance LocalPosition.

The reasons for this design are:  a) both types access data from a pre-captured byte array source so in essence they are the 'same' anyway;  b) it seemed unnecessary to re-implement all the Get* methods simply to separate the method of reading;  and c) it removes ambiguity down the parsing chain so you don't have to inspect the parent ReaderInfo of a Clone and compare its original parsing method.

In the last PR where these classes were actually implemented and working throughout the library, ReaderInfo's are cloned constantly (to achieve the benefit of fewer byte array copies) and passed into other reading/parsing classes outside the parent's control. It's not possible to know down the line which access method a parser will want, and the method can actually change many times depending on the metadata segments. So, it's a better design to have only one type of reader and avoid an extra inspection or clone type change. It also allows for the rare case when indexed and sequential are desirable in a parse.